src/app/berkeley_account_tables/berkeley_account_tables.ml

```diff
@@ -0,0 +1,1644 @@
+  val run :
+       Caqti_async.connection
+    -> state_hash:string
+    -> start_slot:int64
+    -> (int list, [> Caqti_error.call_or_retrieve ]) Deferred.Result.t
+end
+
+let error_count = ref 0
+
+let constraint_constants = Genesis_constants.Constraint_constants.compiled
+
+let proof_level = Genesis_constants.Proof_level.Full
+
+let json_ledger_hash_of_ledger ledger =
+  Ledger_hash.to_yojson @@ Ledger.merkle_root ledger
+
+let create_ledger_as_list ledger =
+  let%map accounts = Ledger.to_list ledger in
+  List.map accounts ~f:(fun acc ->
+      Genesis_ledger_helper.Accounts.Single.of_account acc None )
+
+module First_pass_ledger_hashes = struct
+  (* ledger hashes after 1st pass, indexed by order of occurrence *)
+
+  module T = struct
+    type t = Ledger_hash.Stable.Latest.t * int
+    [@@deriving bin_io_unversioned, compare, sexp, hash]
+  end
+
+  include T
+  include Hashable.Make_binable (T)
+
+  let hash_set = Hash_set.create ()
+
+  let add =
+    let count = ref 0 in
+    fun ledger_hash ->
+      Base.Hash_set.add hash_set (ledger_hash, !count) ;
+      incr count
+
+  let find ledger_hash =
+    Base.Hash_set.find hash_set ~f:(fun (hash, _n) ->
+        Ledger_hash.equal hash ledger_hash )
+
+  (* once we find a snarked ledger hash corresponding to a ledger hash, don't need to store earlier ones *)
+  let flush_older_than ndx =
+    let elts = Base.Hash_set.to_list hash_set in
+    List.iter elts ~f:(fun ((_hash, n) as elt) ->
+        if n < ndx then Base.Hash_set.remove hash_set elt )
+
+  let get_last_snarked_hash, set_last_snarked_hash =
+    let last_snarked_hash = ref Ledger_hash.empty_hash in
+    let getter () = !last_snarked_hash in
+    let setter hash = last_snarked_hash := hash in
+    (getter, setter)
+end
+
+let create_output ~target_epoch_ledgers_state_hash ~target_fork_state_hash
+    ~ledger ~staking_epoch_ledger ~staking_seed ~next_epoch_ledger ~next_seed
+    (input_genesis_ledger : Runtime_config.Ledger.t) =
+  let%bind genesis_ledger_as_list = create_ledger_as_list ledger in
+  let target_genesis_ledger =
+    { input_genesis_ledger with base = Accounts genesis_ledger_as_list }
+  in
+  let%bind staking_epoch_ledger_as_list =
+    create_ledger_as_list staking_epoch_ledger
+  in
+  let%map next_epoch_ledger_as_list = create_ledger_as_list next_epoch_ledger in
+  let target_staking_epoch_data : Runtime_config.Epoch_data.Data.t =
+    let ledger =
+      { input_genesis_ledger with base = Accounts staking_epoch_ledger_as_list }
+    in
+    { ledger; seed = staking_seed }
+  in
+  let target_next_epoch_data : Runtime_config.Epoch_data.Data.t =
+    let ledger =
+      { input_genesis_ledger with base = Accounts next_epoch_ledger_as_list }
+    in
+    { ledger; seed = next_seed }
+  in
+  let target_epoch_data : Runtime_config.Epoch_data.t =
+    { staking = target_staking_epoch_data; next = Some target_next_epoch_data }
+  in
+  { target_fork_state_hash
+  ; target_epoch_ledgers_state_hash
```

Comment:
What is this target epoch ledgers state hash?

Reply:
This code here is a modified replayer. The replayer can use that hash to know when to finish replaying. If there's an output file specified, the epoch ledger with that hash is included, and the hash itself is included. 

It's irrelevant for this app, really, but I didn't want to do too much refactoring.